File: benchmarks/ts/bench.ts
// TypeScript binding benchmark for #86 — binding-layer overhead measurement.
//
// Drives the graphqlite TypeScript binding (bindings/typescript/src) through a
// fixed set of operations against a fresh in-memory graph, timing each op with
// performance.now(). Emits a single JSON document on stdout so benchmarks/run.sh
// can collect it alongside the Python and Rust legs.
//
// IMPORTANT: all three language benchmarks share the *same* C core
// (src/extension.c) via the same staged dylib (GRAPHQLITE_EXTENSION_PATH). These
// numbers are NOT core algorithm speed — they measure the per-binding overhead
// layered on top of that shared core: bulk-insert strategy, value
// escaping/marshalling, per-call round-trips, and result parsing.
//
// Ops (each: 1 warm-up run discarded, then --repeats measured runs, fresh
// :memory: graph per run): build / lookup / scan / pagerank / bfs.
//
// Run via node >=24 (native TS type-stripping + node:sqlite), same as the parity
// runner. The extension is resolved by the binding through GRAPHQLITE_EXTENSION_PATH.
//
//   node bench.ts [--nodes 5000] [--edges 10000] [--repeats 3] [--lookups 2000]

import { arch, platform, release } from 'node:os';
import {
  graph,
  type Graph,
  insertGraphBulk,
  type BulkNodeItem,
  type BulkEdgeItem,
} from '../../bindings/typescript/src/index.ts';

// Deterministic dataset multiplier — shared across the Python/TS/Rust runners so
// every binding builds the byte-identical graph.
const EDGE_K = 3;

interface OpResult {
  name: string;
  ms_median: number;
  ms_min: number;
  runs: number;
  count: number;
}

function parseArgs(argv: string[]): {
  nodes: number;
  edges: number;
  repeats: number;
  lookups: number;
} {
  const opts = { nodes: 5000, edges: 10000, repeats: 3, lookups: 2000 };
  for (let i = 0; i < argv.length; i++) {
    const key = argv[i];
    const val = argv[i + 1];
    if (key === '--nodes' && val !== undefined) (opts.nodes = Number(val)), i++;
    else if (key === '--edges' && val !== undefined) (opts.edges = Number(val)), i++;
    else if (key === '--repeats' && val !== undefined) (opts.repeats = Number(val)), i++;
    else if (key === '--lookups' && val !== undefined) (opts.lookups = Number(val)), i++;
  }
  return opts;
}

function buildDataset(nNodes: number, nEdges: number): {
  nodes: BulkNodeItem[];
  edges: BulkEdgeItem[];
} {
  const nodes: BulkNodeItem[] = new Array(nNodes);
  for (let i = 0; i < nNodes; i++) {
    nodes[i] = [`n${i}`, { val: i }, 'N'];
  }
  const edges: BulkEdgeItem[] = new Array(nEdges);
  for (let i = 0; i < nEdges; i++) {
    edges[i] = [`n${i % nNodes}`, `n${(i * EDGE_K + 1) % nNodes}`, {}, 'E'];
  }
  return { nodes, edges };
}

function median(values: number[]): number {
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0 ? (sorted[mid - 1]! + sorted[mid]!) / 2 : sorted[mid]!;
}

function round4(x: number): number {
  return Math.round(x * 1e4) / 1e4;
}

// Run `setup` (untimed) then time `run` for 1 warm-up + `repeats` runs. `run`
// returns a "count" (result size / op count) used as a cross-language sanity
// check. The warm-up run (r === 0) is discarded.
function measure(
  name: string,
  repeats: number,
  setup: () => Graph,
  run: (g: Graph) => number,
  teardown: (g: Graph) => void,
): OpResult {
  const times: number[] = [];
  let count = 0;
  for (let r = 0; r <= repeats; r++) {
    const ctx = setup();
    const t0 = performance.now();
    count = run(ctx);
    const t1 = performance.now();
    teardown(ctx);
    if (r > 0) times.push(t1 - t0);
  }
  return {
    name,
    ms_median: round4(median(times)),
    ms_min: round4(Math.min(...times)),
    runs: times.length,
    count,
  };
}

function main(): void {
  const { nodes: nNodes, edges: nEdges, repeats, lookups: nLookups } = parseArgs(
    process.argv.slice(2),
  );
  const { nodes, edges } = buildDataset(nNodes, nEdges);
  const lookupIds: string[] = new Array(nLookups);
  for (let i = 0; i < nLookups; i++) lookupIds[i] = `n${i % nNodes}`;
  const seed = 'n0';

  const fresh = (): Graph => graph(':memory:');
  const loaded = (): Graph => {
    const g = graph(':memory:');
    insertGraphBulk(g.connection, nodes, edges);
    return g;
  };
  const close = (g: Graph): void => g.close();

  const ops: OpResult[] = [];

  // build — bulk insert path (fresh empty graph, time only the insert)
  ops.push(
    measure(
      'build',
      repeats,
      fresh,
      (g) => insertGraphBulk(g.connection, nodes, edges).nodesInserted,
      close,
    ),
  );

  // lookup — L individual getNode round-trips (total time)
  ops.push(
    measure(
      'lookup',
      repeats,
      loaded,
      (g) => {
        let hits = 0;
        for (const id of lookupIds) {
          if (g.getNode(id) !== null) hits++;
        }
        return hits;
      },
      close,
    ),
  );

  // scan — full result marshalling
  ops.push(measure('scan', repeats, loaded, (g) => g.query('MATCH (n) RETURN n.id').length, close));

  // pagerank
  ops.push(measure('pagerank', repeats, loaded, (g) => g.pagerank(0.85, 20).length, close));

  // bfs
  ops.push(measure('bfs', repeats, loaded, (g) => g.bfs(seed).length, close));

  const out = {
    binding: 'typescript',
    env: {
      node: process.version,
      v8: process.versions.v8,
      platform: `${platform()}-${release()}`,
      machine: arch(),
    },
    params: { nodes: nNodes, edges: nEdges, repeats, lookups: nLookups },
    ops,
  };
  process.stdout.write(JSON.stringify(out) + '\n');
}

main();

File: benchmarks/python/bench.py
#!/usr/bin/env python3
"""Python binding benchmark for #86 — binding-layer overhead measurement.

Drives the graphqlite Python binding through a fixed set of operations against a
fresh in-memory graph, timing each op with ``time.perf_counter``. Emits a single
JSON document on stdout so ``benchmarks/run.sh`` can collect it alongside the TS
and Rust legs and render a comparison table.

IMPORTANT: all three language benchmarks share the *same* C core
(``src/extension.c``) via the same staged dylib (GRAPHQLITE_EXTENSION_PATH). So
these numbers are NOT a measure of core algorithm speed — they measure the
per-binding overhead layered on top of that shared core: bulk-insert strategy,
value escaping/marshalling, per-call round-trips, and result parsing.

Ops (each: 1 warm-up run discarded, then --repeats measured runs, fresh
``:memory:`` graph per run):

  build     insert_graph_bulk of N nodes + E edges (bulk path)
  lookup    get_node x L individual round-trips (default 2000)
  scan      query("MATCH (n) RETURN n.id") returning all rows (result marshalling)
  pagerank  pagerank(0.85, 20)
  bfs       bfs(seed) full traversal

Usage:
  python bench.py [--nodes 5000] [--edges 10000] [--repeats 3] [--lookups 2000]
"""
from __future__ import annotations

import argparse
import json
import platform
import statistics
import sys
import time
from typing import Any, Callable

import graphqlite

# Deterministic dataset multiplier — shared across the Python/TS/Rust runners so
# every binding builds the byte-identical graph.
EDGE_K = 3


def build_dataset(n_nodes: int, n_edges: int):
    """Deterministic (index-seeded) node/edge lists, identical across bindings.

    node id  = "n{i}",           props = {"val": i} (int), label "N"
    edge i   = ("n{i%N}" -> "n{(i*K+1)%N}"), rel "E", no props
    """
    nodes = [(f"n{i}", {"val": i}, "N") for i in range(n_nodes)]
    edges = [
        (f"n{i % n_nodes}", f"n{(i * EDGE_K + 1) % n_nodes}", {}, "E")
        for i in range(n_edges)
    ]
    return nodes, edges


def _median_min(times_ms: list[float]) -> tuple[float, float]:
    return statistics.median(times_ms), min(times_ms)


def measure(
    name: str,
    repeats: int,
    setup: Callable[[], Any],
    run: Callable[[Any], int],
    teardown: Callable[[Any], None],
) -> dict[str, Any]:
    """Run ``setup`` (untimed) then time ``run`` for 1 warm-up + ``repeats`` runs.

    ``run`` returns an integer "count" (result size / op count) used both as a
    cross-language sanity check and for reporting. The warm-up run is discarded.
    """
    times_ms: list[float] = []
    count = 0
    for r in range(repeats + 1):
        ctx = setup()
        t0 = time.perf_counter()
        count = run(ctx)
        t1 = time.perf_counter()
        teardown(ctx)
        if r > 0:  # discard warm-up (r == 0)
            times_ms.append((t1 - t0) * 1000.0)
    median_ms, min_ms = _median_min(times_ms)
    return {
        "name": name,
        "ms_median": round(median_ms, 4),
        "ms_min": round(min_ms, 4),
        "runs": len(times_ms),
        "count": count,
    }


def main() -> int:
    ap = argparse.ArgumentParser(description="graphqlite Python binding benchmark")
    ap.add_argument("--nodes", type=int, default=5000)
    ap.add_argument("--edges", type=int, default=10000)
    ap.add_argument("--repeats", type=int, default=3)
    ap.add_argument("--lookups", type=int, default=2000)
    args = ap.parse_args()

    n_nodes, n_edges = args.nodes, args.edges
    repeats, n_lookups = args.repeats, args.lookups
    nodes, edges = build_dataset(n_nodes, n_edges)
    lookup_ids = [f"n{i % n_nodes}" for i in range(n_lookups)]
    seed = "n0"

    def fresh():
        return graphqlite.graph(":memory:")

    def loaded():
        g = graphqlite.graph(":memory:")
        g.insert_graph_bulk(nodes, edges)
        return g

    def close(g):
        g.close()

    ops = []

    # build — bulk insert path (fresh empty graph, time only the insert)
    ops.append(
        measure(
            "build",
            repeats,
            setup=fresh,
            run=lambda g: g.insert_graph_bulk(nodes, edges).nodes_inserted,
            teardown=close,
        )
    )

    # lookup — L individual get_node round-trips (total time)
    def run_lookup(g) -> int:
        hits = 0
        for nid in lookup_ids:
            if g.get_node(nid) is not None:
                hits += 1
        return hits

    ops.append(
        measure("lookup", repeats, setup=loaded, run=run_lookup, teardown=close)
    )

    # scan — full result marshalling
    def run_scan(g) -> int:
        rows = g.query("MATCH (n) RETURN n.id")
        return len(rows)

    ops.append(measure("scan", repeats, setup=loaded, run=run_scan, teardown=close))

    # pagerank
    ops.append(
        measure(
            "pagerank",
            repeats,
            setup=loaded,
            run=lambda g: len(g.pagerank(0.85, 20)),
            teardown=close,
        )
    )

    # bfs
    ops.append(
        measure(
            "bfs",
            repeats,
            setup=loaded,
            run=lambda g: len(g.bfs(seed)),
            teardown=close,
        )
    )

    out = {
        "binding": "python",
        "env": {
            "python": platform.python_version(),
            "implementation": sys.implementation.name,
            "platform": platform.platform(),
            "machine": platform.machine(),
        },
        "params": {
            "nodes": n_nodes,
            "edges": n_edges,
            "repeats": repeats,
            "lookups": n_lookups,
        },
        "ops": ops,
    }
    json.dump(out, sys.stdout)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
